### PR TITLE
WIP: Adding support for razor API behind HTTPS

### PR DESCRIPTION
**Pull request by Nekototori**
I modified the rest call to add supported HTTPS methods along with a way to specify in the API certs etc using the existing contructs you provided.

This allows anyone running razor with an ssl cert in front of the API to use this module.

### CODE REVIEW

File: lib/puppet/provider/razor_rest.rb [hunk 1]
```diff
@@ -33,11 +33,38 @@ def self.get_rest_info
     else
       port = 8150
     end
+     if yamldata.include?('http_method')
```
Comment (scottsolowild) [inline lib/puppet/provider/razor_rest.rb]: Newline + indentation.

File: lib/puppet/provider/razor_rest.rb [hunk 2]
```diff
@@ -110,10 +152,10 @@ def self.get_json_from_url(url)
...
     rest = get_rest_info
-    url = "http://#{rest[:ip]}:#{rest[:port]}/api"    
+    url = "#{rest[:http]}://#{rest[:ip]}:#{rest[:port]}/api"    
     
     responseJson = get_json_from_url(url)
     version = responseJson["version"]["server"] || "Unknown"            
     version
   end
-end
+end
```
Comment (scottsolowild) [inline lib/puppet/provider/razor_rest.rb]: Newline here disappeared.

File: manifests/api.pp [hunk 3]
```diff
@@ -16,6 +16,10 @@
 class razor::api (
   String $hostname = 'localhost',
   Variant[Undef, Integer] $port = undef,
+  Variant[String, Enum['http', 'https']] $http_method = 'http',
```
Comment (scottsolowild) [inline manifests/api.pp]: These should be added to the documentation above.

File: lib/puppet/provider/razor_rest.rb [hunk 4]
```diff
@@ -81,10 +108,17 @@ def post_command(command, resourceHash)
...
+    ssl_rest = RestClient::Resource.new(
+        url,
+        :ssl_client_cert => OpenSSL::X509::Certificate.new(File.read("#{rest[:client_cert]}")),
+        :ssl_client_key  => OpenSSL::PKey::RSA.new(File.read("#{rest[:private_key]}")),
+        :ssl_ca_file     => "#{rest[:ca_cert]}",
+)
+
     begin
-      RestClient.post url, resourceHash.to_json, :content_type => :json
+#      RestClient.post url, resourceHash.to_json, :content_type => :json
```
Comment (scottsolowild) [inline lib/puppet/provider/razor_rest.rb]: Need this?

File: lib/puppet/provider/razor_rest.rb [hunk 5]
```diff
@@ -33,11 +33,38 @@ def self.get_rest_info
...
+     if yamldata.include?('http_method')
+      http = yamldata['http_method']
+    else
+      http = 'http'
+    end
+
+    if yamldata.include?('client_cert')
+      client_cert = yamldata['client_cert']
+    else
+      client_cert = "/etc/puppetlabs/puppet/ssl/certs/#{hostname}.pem"
```
Comment (scottsolowild) [inline lib/puppet/provider/razor_rest.rb:46]: I think these paths should default to `nil`. The installation module should be responsible for helping this module locate these files. I suggest adding these as module parameters in `init.pp`.

**Review by scottsolowild — CHANGES_REQUESTED** [review]

File: manifests/api.pp (hunk 3 above)
Comment (natemccurdy) [reply manifests/api.pp]: @Nekototori Should this parameter be only one of `http` and `https`?

If so, you don't need to Variant String stuff, and can do just `Enum['http', 'https']`.

File: lib/puppet/provider/razor_rest.rb (hunk 5 above)
Comment (Nekototori) [reply lib/puppet/provider/razor_rest.rb:46]: Mostly just following the existing code logic to maintain consistency.

**Pull request merged**